x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,432 @@
+            QueryShardContext context
+        ) {
+            if (context.allowExpensiveQueries() == false) {
+                throw new ElasticsearchException(
+                    "[fuzzy] queries cannot be executed when '" + ALLOW_EXPENSIVE_QUERIES.getKey() + "' is set to false."
+                );
+            }
+            failIfNotIndexed();
+            return new FuzzyQuery(
+                new Term(name(), (BytesRef) value),
+                fuzziness.asDistance(BytesRefs.toString(value)),
+                prefixLength,
+                maxExpansions,
+                transpositions
+            ) {
+                @Override
+                protected TermsEnum getTermsEnum(Terms terms, AttributeSource atts) throws IOException {
+                    ByteRunAutomaton runAutomaton = getAutomata().runAutomaton;
+
+                    return new FilteredTermsEnum(terms.iterator(), false) {
+
+                        @Override
+                        protected AcceptStatus accept(BytesRef term) throws IOException {
+                            byte[] decoded = VersionEncoder.decodeVersion(term).getBytes(StandardCharsets.UTF_8);
+                            boolean accepted = runAutomaton.run(decoded, 0, decoded.length);
+                            if (accepted) {
+                                return AcceptStatus.YES;
+                            }
+                            return AcceptStatus.NO;
+                        }
+                    };
+                }
+            };
+        }
+
+        @Override
+        public Query wildcardQuery(String value, MultiTermQuery.RewriteMethod method, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
+                throw new ElasticsearchException(
+                    "[wildcard] queries cannot be executed when '" + ALLOW_EXPENSIVE_QUERIES.getKey() + "' is set to false."
+                );
+            }
+            failIfNotIndexed();
+
+            VersionFieldWildcardQuery query = new VersionFieldWildcardQuery(new Term(name(), value));
+            QueryParsers.setRewriteMethod(query, method);
+            return query;
+        }
+
+        @Override
+        protected BytesRef indexedValueForSearch(Object value) {
+            String valueAsString;
+            if (value instanceof String) {
+                valueAsString = (String) value;
+            } else if (value instanceof BytesRef) {
+                // encoded string, need to re-encode
+                valueAsString = ((BytesRef) value).utf8ToString();
+            } else {
+                throw new IllegalArgumentException("Illegal value type: " + value.getClass() + ", value: " + value);
+            }
+            return encodeVersion(valueAsString).bytesRef;
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName, Supplier<SearchLookup> searchLookup) {
+            failIfNoDocValues();
```

Comment:
if we decide to always index a field with doc values, this `failIfNoDocValues()` check seems unnecessary.